### PR TITLE
Improve ssl settings

### PR DESCRIPTION
**Pull request by Mr-Pi**
* Add Diffie-Hellman parameter
* Add OCSP Stapling
* Better default cipher selection

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,19 +14,32 @@ A Debian based distribution with certbot available in current apt sources. Corre
...
+| Option                                  | Type            | Default                                          | Description                                                         |          Required         |
+|-----------------------------------------|-----------------|--------------------------------------------------|---------------------------------------------------------------------|:-------------------------:|
+| proxy_domains                           | list of dicts   |                                                  | List of all target servers                                          |             Y             |
+| default_url                             | string          | `https://github.com/stuvusIT/reverse_proxy`      | Url to redirect to if no target with requested domain is configured |             N             |
+| letsencrypt_email                       | string          | `false`                                          | E-Mail address to use to request certificates                       |             Y             |
```
Comment (SlothOfAnarchy) [inline README.md]: Why is this required if there is a default set?

File: README.md [hunk 2]
```diff
@@ -14,19 +14,32 @@ A Debian based distribution with certbot available in current apt sources. Corre
...
+| reverse_proxy_dhparam_path              | string          | `/etc/ssl/dhparam.pem`                           | Path to store the Diffie-Hellman parameters                         |             N             |
+| reverse_proxy_ssl_session_timeout       | string          | `1d`                                             | nginx `ssl_session_timeout` option                                  |             N             |
+| reverse_proxy_ssl_session_cache         | string          | `shared:SSL:50m`                                 | nginx `ssl_session_cache` option                                    |             N             |
+| reverse_proxy_ssl_session_tickets       | boolean         | `False`                                          | nginx `ssl_session_tickets` option                                  |             N             |
+| reverse_proxy_ssl_protocols             | string          | `TLSv1 TLSv1.1 TLSv1.2`                          | nginx `ssl_protocols` option                                        |             N             |
```
Comment (SlothOfAnarchy) [inline README.md]: Please use a list for this and join its contents with `|join(" ")`

File: README.md [hunk 3]
```diff
@@ -14,19 +14,32 @@ A Debian based distribution with certbot available in current apt sources. Corre
...
+| reverse_proxy_ssl_ciphers               | string          | ___see [defaults/main.yml](defaults/main.yml)___ | nginx `ssl_ciphers` option                                          |             N             |
```
Comment (SlothOfAnarchy) [inline README.md]: Please use a list for this and join its contents with `|join(":")`. One can decipher the used ciphers much easier.

File: README.md [hunk 4]
```diff
@@ -14,19 +14,32 @@ A Debian based distribution with certbot available in current apt sources. Corre
...
+| reverse_proxy_ssl_prefer_server_ciphers | boolean         | True                                             | nginx `ssl_prefer_server_ciphers` option                            |             N             |
```
Comment (SlothOfAnarchy) [inline README.md]: use `block quotes` for these two default values.

File: tasks/dhparam.yml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+---
+- name: Get age of Diffie-Hellman parameters
+  stat:
+    path: "{{ reverse_proxy_dhparam_path }}"
+  register: reverse_proxy_dhparam_stats
+  changed_when: False
+
+- name: Generate Diffie-Hellman parameters if needed
+  command: openssl dhparam -out {{ reverse_proxy_dhparam_path }} {{ reverse_proxy_dhparam_size }}
+  when: reverse_proxy_dhparam_stats.stat.exists == False or reverse_proxy_dhparam_stats.stat.mtime|int + reverse_proxy_dhparam_max_age|int >= ansible_date_time.epoch
```
Comment (SlothOfAnarchy) [inline tasks/dhparam.yml]: I'd have used `not reverse_proxy_dhparam_stats.stat.exists` but you don't have to change it.

File: tasks/main.yml [hunk 6]
```diff
@@ -4,4 +4,6 @@
   with_items: "{{ proxy_domains }}"
   loop_control:
     loop_var: target
-- import_tasks: nginx.yml
\ No newline at end of file
+- import_tasks: dhparam.yml
+  when: reverse_proxy_use_dhparam == True
```
Comment (SlothOfAnarchy) [inline tasks/main.yml]: ` == True` is redundant

File: templates/reverse_proxy.conf.j2 [hunk 7]
```diff
@@ -17,13 +17,33 @@ server {
 
 	ssl_certificate     /etc/letsencrypt/live/{{- build_domain_list_for_domains( domains, ' ' ).split(' ')[1] -}}/fullchain.pem;
 	ssl_certificate_key /etc/letsencrypt/live/{{- build_domain_list_for_domains( domains, ' ' ).split(' ')[1] -}}/privkey.pem;
+	ssl_session_timeout {{ reverse_proxy_ssl_session_timeout }};
+	ssl_session_cache {{ reverse_proxy_ssl_session_cache }};
+	ssl_session_tickets {% if reverse_proxy_ssl_session_tickets == True %}on{% else %}off{% endif %};
```
Comment (SlothOfAnarchy) [inline templates/reverse_proxy.conf.j2]: ` == True` is redundant

File: templates/reverse_proxy.conf.j2 [hunk 8]
```diff
@@ -17,13 +17,33 @@ server {
...
 
+	ssl_protocols {{ reverse_proxy_ssl_protocols }};
+	ssl_ciphers {{ reverse_proxy_ssl_ciphers }};
+	ssl_prefer_server_ciphers {% if reverse_proxy_ssl_prefer_server_ciphers == True %}on{% else %}off{% endif %};
```
Comment (SlothOfAnarchy) [inline templates/reverse_proxy.conf.j2]: ` == True` is redundant

File: templates/reverse_proxy.conf.j2 [hunk 9]
```diff
@@ -17,13 +17,33 @@ server {
...
+
+	add_header Strict-Transport-Security "max-age={{ hsts_max_age }}; includeSubDomains; preload" always;
+
+{% if reverse_proxy_ssl_stapling == True %}
```
Comment (SlothOfAnarchy) [inline templates/reverse_proxy.conf.j2]: ` == True` is redundant

File: templates/reverse_proxy.conf.j2 [hunk 10]
```diff
@@ -17,13 +17,33 @@ server {
...
+	ssl_stapling on;
+	ssl_stapling_verify on;
+	{% if reverse_proxy_ssl_trusted_certificate is defined -%}
+	ssl_trusted_certificate {{ reverse_proxy_ssl_trusted_certificate }};
+	{%- endif %}
+
+{% endif %}
+{% if reverse_proxy_use_dhparam == True %}
```
Comment (SlothOfAnarchy) [inline templates/reverse_proxy.conf.j2]: ` == True` is redundant

File: README.md [hunk 11]
```diff
@@ -14,19 +14,32 @@ A Debian based distribution with certbot available in current apt sources. Corre
...
+| reverse_proxy_ssl_stapling              | boolean         | True                                             | Enable OCSP Stapling                                                |             N             |
+| reverse_proxy_ssl_trusted_certificate   | string          |                                                  | nginx ssl_trusted_certificate option                                |             N             |
```
Comment (SlothOfAnarchy) [inline README.md]: This is a path, right? If so, please reflect this in the description.

File: README.md (hunk 2 above)
Comment (Mr-Pi) [reply README.md]: Good point

File: README.md (hunk 3 above)
Comment (Mr-Pi) [reply README.md]: Good point

File: templates/reverse_proxy.conf.j2 (hunk 7 above)
Comment (Mr-Pi) [reply templates/reverse_proxy.conf.j2]: I believe, there is no default?

Comment (SlothOfAnarchy) [reply templates/reverse_proxy.conf.j2]: You don't need a default. What I mean is that if will already check the content of the variable, you don't need to explicitly say it should be `True`.

**Comment by Mr-Pi** [comment]
Please rereview

File: templates/reverse_proxy.conf.j2 (hunk 7 above)
Comment (Mr-Pi) [reply templates/reverse_proxy.conf.j2]: I'm not sure, but I will test it

Comment (Mr-Pi) [reply templates/reverse_proxy.conf.j2]: Seems good, I changed it.
